cli/explain/compactgraph/testdata/generate.go

```diff
@@ -0,0 +1,224 @@
+		changes      string
+		changedArgs  []string
+	}{
+		{
+			name:     "java_noop_impl_change",
+			baseline: JavaProject,
+			changes: `
+-- src/main/java/com/example/lib/Lib.java --
+package com.example.lib;
+
+public class Lib {
+    public static String getName() {
+      return "Lib"; // no-op impl change
+    }
+}
+`,
+		},
+		{
+			name:     "java_impl_change",
+			baseline: JavaProject,
+			changes: `
+-- src/main/java/com/example/lib/Lib.java --
+package com.example.lib;
+
+public class Lib {
+    public static String getName() {
+      return "Lib2"; // impl change
+    }
+}
+`,
+		},
+		{
+			name:     "java_header_change",
+			baseline: JavaProject,
+			changes: `
+-- src/main/java/com/example/lib/Lib.java --
+package com.example.lib;
+
+public class Lib {
+    public static String getName() {
+      return "Lib";
+    }
+
+    public static void foo() {}
+}
+`,
+		},
+		{
+			name: "env_change",
+			baseline: `
+-- MODULE.bazel --
+-- pkg/BUILD --
+genrule(
+    name = "gen",
+	outs = ["out"],
+	cmd = "env > $@",
+)
+`,
+			baselineArgs: []string{"--action_env=EXTRA=foo", "--action_env=OLD_ONLY=old_only", "--action_env=OLD_AND_NEW=old"},
+			changedArgs:  []string{"--action_env=NEW_ONLY=new_only", "--action_env=OLD_AND_NEW=new", "--action_env=EXTRA=foo"},
+		},
+		{
+			name: "non_hermetic",
+			baseline: `
+-- MODULE.bazel --
+-- pkg/BUILD --
+genrule(
+    name = "gen",
+	outs = ["out"],
+	cmd = "uuidgen > $@",
+)
+`,
+		},
+	} {
+		tmpDir, err := os.MkdirTemp("", "explain-test-*")
+		if err != nil {
+			log.Fatalf("Failed to create temp dir: %s", err)
+		}
+		defer os.RemoveAll(tmpDir)
+
+		extractTxtar(tmpDir, tc.baseline)
+		collectLog(tc.baselineArgs, tmpDir, filepath.Join(outDir, tc.name+"_old.pb.zstd"))
+
+		extractTxtar(tmpDir, tc.changes)
+		collectLog(tc.changedArgs, tmpDir, filepath.Join(outDir, tc.name+"_new.pb.zstd"))
+	}
+}
+
+func collectLog(args []string, projectDir, logPath string) {
+	outputBase, err := os.MkdirTemp("", "explain-testdata-*")
+	if err != nil {
+		log.Fatalf("Failed to create temp output base: %s", err)
+	}
+	defer os.RemoveAll(outputBase)
+	// Bazel's output base can contain files with no write permissions.
+	defer filepath.WalkDir(outputBase, func(path string, d fs.DirEntry, err error) error {
+		return os.Chmod(path, 0755)
+	})
+	cmd := exec.Command(
+		"bazel",
+		"--nohome_rc", "--nosystem_rc",
+		"--output_base="+outputBase,
+		"test", "//...",
+	)
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Args = append(
+		cmd.Args,
+		"--java_runtime_version=remotejdk_21",
+		"--experimental_execution_log_compact_file="+logPath,
+	)
+	cmd.Dir = projectDir
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err = cmd.Run(); err != nil {
+		// Allow failures due to no tests as we always run with `bazel test`.
+		if exitErr, ok := err.(*exec.ExitError); !ok || exitErr.ExitCode() != 4 {
+			log.Fatalf("Failed to run command: %s", err)
+		}
+	}
+}
+
+func extractTxtar(dir string, tar string) {
```

Comment:
TIL about txtar - looks useful :)

Reply:
Go 1.23 even has `os.CopyFS` in the stdlib, but unfortunately we can't use it here (it errors out when a file already exists).